lib/Index/Index.cpp

```diff
@@ -440,6 +464,270 @@ class IndexSwiftASTWalker : public SourceEntityWalker {
+  /// stringValue)' we can see that 'a' corresponds to 'Int' and 'intValue',
+  /// while 'b' corresponds to 'String' and 'stringValue'. By identifying these
+  /// relationships, we can pinpoint locations within the PatternBindingDecl
+  /// that the corresponding VarDecl should be marked as the current active
+  /// container. For example, when we walk to the TypeRepr for 'Int', we can
+  /// make 'a' the current active container, and likewise for 'String', 'b'. And
+  /// thus, when the walk reaches the point of creating the reference for 'Int'
+  /// it will be contained by 'a'.
+  ///
+  /// These locations are also identified for single named pattern bindings; in
+  /// which case, the VarDecl is activated for all types and expressions with
+  /// the pattern.
+  ///
+  /// Pattern bindings containing an AntPattern (i.e let _) are a special case,
+  /// as they have no VarDecl. Their types and expressions are associated with
+  /// the current active container, if any. Therefore, given such a pattern
+  /// declared within a function, the type and expression references will be
+  /// contained by the function. If there is no active container, the references
+  /// are not contained.
+  void identifyContainerVarDecls(const PatternBindingDecl *const bindingDecl,
+                                 ContainerStackEntry &entry) const {
+    if (auto varDecl = bindingDecl->getSingleVar()) {
+      // This is a single var binding; therefore, it may also have custom
+      // attributes. Immediately activate the VarDecl so that it can be the
+      // container for the attribute references.
+      entry.activeDecls = {varDecl};
+    }
+
+    for (const auto patternIndex : range(bindingDecl->getNumPatternEntries())) {
+      auto pattern = bindingDecl->getPattern(patternIndex);
+      PatternVarDeclMap varDeclMap;
+      collectPatternVarDecls(pattern, varDeclMap);
+
+      if (varDeclMap.size() == 1) {
+        auto optionalVarDecl = varDeclMap.begin()->second;
+        setupContainerStackEntry(entry, pattern, optionalVarDecl);
+        continue;
+      }
+
+      if (auto typedPattern = dyn_cast<TypedPattern>(pattern)) {
+        auto typeRepr = typedPattern->getTypeRepr();
+
+        if (auto tupleTypeRepr = dyn_cast<TupleTypeRepr>(typeRepr)) {
+          associatePatternTypes(tupleTypeRepr, varDeclMap, entry);
+        } else {
+          // Non-tuple type, associate all VarDecls in this pattern.
+          entry.declMap[typeRepr] = {false, pattern};
+        }
+      }
+
+      if (auto expr = bindingDecl->getInit(patternIndex)) {
+        if (auto tupleExpr = dyn_cast<TupleExpr>(expr)) {
+          associatePatternExpressions(tupleExpr, varDeclMap, entry);
+        } else {
+          // Non-tuple initializer, associate all VarDecls in this pattern.
+          entry.declMap[expr] = {false, pattern};
+        }
+      }
+    }
+  }
+
+  void collectPatternVarDecls(const Pattern *const pattern,
```

Comment:
Interesting idea with the `path`. It'd be nice to avoid all the string manipulation and copying though.

I haven't thought this through completely, but since we have a guaranteed order here, could we instead construct a vector of {`PointerUnion<Decl *, Pattern*>` and number of children} in the initial collection?

`let (a, ((b, c), d), e)` ->
```
[
  Decl(a),
  {Pattern(bcd), 4},
    {Pattern(bc), 2},
      Decl(b), Decl(c),
    Decl(d),
  Decl(e)
]
```

Then for associating the type and expression, iterate through the vector and skip the number of children if the type/expression isn't a Tuple. So for `: (Int, (SomeAlias, Int), String) = (1, foo(), "")` ->
Walking types:
  - `Int` -> associate to current index and skip children (0 so no skip), go to next index
  - `Tuple` -> expect a pattern (have to decide what to do if not), enter type and go to next index
  - `Tuple` -> expect pattern, enter type and go to next index
  - `Int` -> associate, skip, go to next index
  - `String -> ditto
  - `Int` -> ditto
  - `String` -> ditto

Walking expression:
  - `1` -> associate, skip, go to next index
  - `foo()` -> associate, skip (4 children in this case), go to next index
  - `""` -> associate, skip, go to next index